### PR TITLE
 [KEYCLOAK-8575] oidc idp basic auth

### PR DESCRIPTION
**Pull request by rradillen**
Allow to choose between basic auth and form auth for OIDC idp.

docs: https://github.com/keycloak/keycloak-documentation/pull/722

### CODE REVIEW

File: server-spi-private/src/main/java/org/keycloak/broker/provider/util/SimpleHttp.java
Comment (mposolda) [inline server-spi-private/src/main/java/org/keycloak/broker/provider/util/SimpleHttp.java]: Please remove star imports - from all java classes in the PR. In Keycloak, we prefer to avoid star imports.

**Review by mposolda — CHANGES_REQUESTED** [review]
Thanks for the contribution! PR looks good to me. Just 2 things:
- Please remove star imports from all java files (added inline comment for it as well)
- Will be good to have the automated test for this switch. See package "org.keycloak.testsuite.broker" .

**Comment by vmuzikar** [comment]
Hi @rradillen! Thank you for your contribution.

@mposolda I just wanted to ask if this change affects the social identity providers. Should we run the SocialLoginTest with it?

**Comment by rradillen** [comment]
> @mposolda I just wanted to ask if this change affects the social identity providers. Should we run the SocialLoginTest with it?

As far as I can tell the Kecyloak, Gitlab and Google identityprovider inherit from the OIDCIdentityProvider instead of the AbstractOAuth2IdentityProvider. The default value is still to retain the auth in post configuration so they should remain working but indeed you might want to test these.

**Comment by mposolda** [comment]
@vmuzikar yes, as @rradillen pointed, the social providers shouldn't be affected. That's the good difference in comparing to the other PR we had few months ago, which changed default behaviour and broke few social providers.

**Comment by vmuzikar** [comment]
@mposolda @rradillen Ok, thank you for the confirmation. Just to be sure I've run the SocialLoginTest and everything passed as expected.

**Comment by mposolda** [comment]
@rradillen The PR looks good. However it will require rebase.

Moreover, we have other PR, which touches OIDC client authentication too - https://github.com/keycloak/keycloak/pull/6303 . So my suggestion would be first wait for the other PR 6303 to be merged. That will include also documentation update, so you won't need to update the documentation as part of your PR.

I will ping again once the other PR is merged and available. The only thing required will be rebase and align your PR with the other PR. The rebase should also fix the travis error in RefreshTokenTest (it was fixed in the meantime in keycloak master).

**Comment by mposolda** [comment]
@rradillen The other PR https://github.com/keycloak/keycloak/pull/6303 was merged. Do you have a chance to rebase and align your PR with the other PR? For example the other PR already added option clientAuthMethod, but with the value clientAuth_basic commented.

**Comment by mposolda** [comment]
@rradillen Thanks for the update. However there are test failures of broker tests (See travis run for more details https://travis-ci.org/keycloak/keycloak/jobs/586169320 ) and I am seeing that there will be more changes needed in your PR.

For example the other PR already introduced the option clientAuthMethod  on OAuth2IdentityProviderConfig. So there is no need to re-introduce the same option in your PR. Just add the support for oidc idp basic auth. Similarly the other PR introduced the changes in AbstractOAuth2IdentityProvider, so you may just need to align with them and add support for OIDC Basic Auth? IMO it is fine to handle changes in AbstractOAuth2IdentityProvider rather than OIDCIdentityProvider as client-auth-basic is not OIDC specific stuff, but can be potentially useful for some social providers in the future.

Sorry for more work for you. 2 community PRs touching same part of the codebase means conflict for one or another and the other PR was faster...

**Comment by rradillen** [comment]
Working on it when I have the time. Can you clarify on how to run the tests locally because waiting for travis is a bit of a long feedback cycle.

**Comment by mposolda** [comment]
@rradillen Cool, thanks. Running the tests for example:
```
mvn clean install -DskipTests=true
cd testsuite/integration-arquillian
mvn clean install -Dtest=*.broker.*
```
or some variant of second command to run only single test:
mvn clean install -Dtest=OidcUserAttributeMapperTest

you can check in travis results, which tests were broken in your PR.

File: themes/src/main/resources/theme/base/admin/messages/admin-messages_en.properties
Comment (mposolda) [inline themes/src/main/resources/theme/base/admin/messages/admin-messages_en.properties]: The stuff was already added to this file in the other PR. I think that in your PR, you don't need to touch this file at all.

File: themes/src/main/resources/theme/base/admin/resources/partials/realm-identity-provider-oidc.html
Comment (mposolda) [inline themes/src/main/resources/theme/base/admin/resources/partials/realm-identity-provider-oidc.html]: Switch was already added and in the other PR, you already uncommented the line allowing clientAuth_basic in the existing switch. That should be sufficient to uncomment that line. No need to add the switch again to have it duplicated on the page. Could you please remove the duplicated switch from your PR?

File: services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProviderConfig.java
Comment (mposolda) [inline services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProviderConfig.java]: Maybe this constant can be removed too, it seems it is not used anywhere now?

File: services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java
Comment (mposolda) [inline services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java]: Bonus point: If you're able to add stuff from AbstractOAuth2IdentityProvider.generateTokenRequest to support JWT authentication here, it will be great! Looks like this capability is missing in the PR, which added the support for JWT authentication. Thanks

**Review by mposolda — CHANGES_REQUESTED** [review]
Thanks for your work. I've added few more review comments as some duplications seem to be still here. Also there is one issue in the original PR, which added support for JWT authentication, that it didn't update some refreshToken calls in OIDCIdentityProvider class. If you manage to fix it in your PR, it will be great. But that's just bonus point, it is not your fault, so we can eventually create follow-up JIRA if needed. Thanks

**Review by mposolda — APPROVED** [review]
@rradillen Thanks

**Pull request merged**
